Dockerfile

```diff
@@ -5,6 +5,20 @@ WORKDIR /code
 
 COPY entrypoint.sh /entrypoint.sh
 COPY node-build.sh /node-build.sh
-RUN chmod +x /entrypoint.sh
-RUN chmod +x /node-build.sh
-CMD [ "bash", "/entrypoint.sh" ]
+
+# set execute and install browsers
+RUN chmod +x /entrypoint.sh && \
+    chmod +x /node-build.sh && \
+    apt-get update && \
+    apt-get install -y chromium firefox-esr
```

Comment:
Few tips to improve this:

- You should always clean up the apt list after installing something 
- You should specify the "--no-install-recommends" flag to limit potentially unwanted software.
- You should use specific version tags on software that you install to make sure that they don't change from build to build.
- Separating the chromium and firefox installs into multiple RUN commands improves reusability of the layer in case either version changes.

I.m.o this should become:

```
Set execute permission on scripts
RUN chmod +x /entrypoint.sh && \
    chmod +x /node-build.sh

RUN apt-get update && \
    apt-get install -y --no-install-recommends chromium=ver && \
	rm -rf /var/lib/apt/lists/*

RUN apt-get update && \
    apt-get install -y --no-install-recommends firefox-esr=ver && \
	rm -rf /var/lib/apt/lists/*
```

I also don't think this code belongs in the node-build container. These are front-end specific dependencies.

I would make a new container which is based on this one with the front-end only stuff.

Reply:
First of all, Thanks for the comments (on the weekends :-) )! 

Items 1.2.3. fixed: good to fix versions for tracability and security (f.e. package injection)

Not sure about item 4, layers are build on top of each other, so if chromium ver changes in this example, then the next layers has to be rebuild as well right? Also both versions change very rapidly, most of the time you want to update both. I think for now better to keep them inline to prevent code duplication maintenance issues and performance as well.

Final items: fine with a new container because of the impact for now, but backend build might depend on sonar in the future as well. And maybe also on browsers running int, e2e or unittest...

Reply:
We are talking about different caches :smile: 

Build caching only checks whether the command has changed. Not whether any checksums have changed. If you **do** change one and not the other it will rebuild both if you don't split them.

Read-caches only check the checksum of a layer.

Layers are unique (/independent) unless they modify the same path/file, they can only do that whilst in the same writing layer (e.g container layer).

So pulling a container with a different firefox version and keeping chrome the same would only update a single layer.

For builds, it isn't technically true that layers need rebuilding once a lower layer changes though. Especially with the RUN command the build cache only compares the Dockerfile statement (as opposed to file-checksums as it does with ADD/COPY).
The only layer that will always show change is the very latest one (as that is where the **actual** container build happens)

I don't agree that splitting firefox and chromium can be counted as code duplication. The only reason I'm duplicating the update/removal of the apt lists is to make each layer smaller. (to that effect we should even add `apt clean`).

For me, the added benefits (mainly caching and security) outweigh the cons (e.g a few duplicate lines). To identify where I split my run commands I usually think of doing it multi-staged and split according to that. (in which case the separation is a given)



PS: because you added firefox-esr that won't nearly change as much as the non-long-term releases of chromium. I don't think it makes for a valid counter-argument though as we probably won't keep up with releases anyway so I just wanted to mention it in the PS :smile: 

(note that the esr version number is also different than the version number of a dev/user browser)

Reply:
Interesting arguments i will do some further testing on it

Reply:
Feel free! I have a very detailed (1.5 hour long) unlisted conference video from a couple of years ago where I dive pretty deep into layering and UnionFS :smile: